virtwho/config.py

```diff
@@ -819,18 +828,62 @@ def __len__(self):
     def __getitem__(self, key):
         return self._values[key]
 
-    def validate(self):
-        validation_messages = []
-        # Do validation in subclasses
-        self.validation_messages = validation_messages
-        if len(self._values) == 0:
-            validation_messages.append(('warning', 'No values provided'))
-
+    def _pre_validate(self):
+        """
+        Steps necessary to do before evaluation 
+        """
+        # FIXME: add some comments with some explanation
         if 'virttype' in self._values:
             self._values['type'] = self._values['virttype']
+
+        if len(self._unvalidated_keys) == 0:
+            self.validation_messages.append(('warning', 'No values provided in: %s' % self.name))
+        else:
+            for default_key in self.defaults.keys():
+                if default_key not in self._unvalidated_keys:
+                    self.validation_messages.append(
+                        ('warning', 'Value for %s not set in: %s, using default: %s' %
+                         (default_key, self.name, self.defaults[default_key]))
+                    )
+
+    def _post_validate(self):
+        """
+        Steps necessary to do after evaluation
+        """
+        for required_key in self.REQUIRED:
+            if required_key not in self:
+                msg = ('error', 'Required option: "%s" is missing in: %s' % (required_key, self.name))
+                self.validation_messages.append(msg)
+        self.reset_to_defaults()
         # Finally calls _update_state
         self._update_state()
-        return validation_messages
+
+    def _validate(self):
+        """
+        Own implementation should be done in subclasses
+        """
+        # raise RuntimeError("This method should never be called and should be implemented in subclass")
```

Comment:
The instance of `ConfigSection` is currently used and uncommenting this would completely break virt-who ATM. I guess it is going to be changed soon and we will use only subclasses of `ConfigSection`. Then it will be possible uncomment this line, right? @cnsnyder